proposals/ZZZZ-identity-keypath.md

```diff
@@ -0,0 +1,112 @@
+# Identity key path
+
+* Proposal: [SE-ZZZZ](ZZZZ-identity-keypath.md)
+* Authors: [Joe Groff](https://github.com/jckarter)
+* Review Manager: TBD
+* Status: **Awaiting review**
+* Implementation: [apple/swift#18804](https://github.com/apple/swift/pull/18804)
+
+## Introduction
+
+Add the ability to reference the identity key path, which refers to the entire
+input value it is applied to.
+
+Swift-evolution thread: [Some small keypath extensions: identity and tuple components](https://forums.swift.org/t/some-small-keypath-extensions-identity-and-tuple-components/13729)
+
+## Motivation
+
+Key paths provide a means to refer to part of a value or a path through an
+object graph independent of any specific instance. In most places where this
+is useful, it is also useful to be able to refer to the entire value.
+For instance, one could have a coordinator object that owns a state value and
+notifies observers of changes to the state by allowing modification through
+key paths:
+
+```swift
+class ValueController<T> {
+  private var state: T
+
+  private var observers: [(T) -> ()]
+
+  subscript<U>(key: WritableKeyPath<T, U>) {
+    get { return state[keyPath: key] }
+    set {
+      state[keyPath: key] = newValue
+      for observer in observers {
+        observer(state)
+      }
+    }
+  }
+}
+```
+
+With such an interface, it'd be useful to be able to update the entire state
+object at once.
+
+## Proposed solution
+
+We add a way to refer to the **identity key path**, which refers to the entire
+input value a key path applies to.
+
+## Detailed design
+
+Every value in Swift has a special pseudo-property `.self`, which refers to
+the entire value:
+
+```swift
+var x = 1
+x.self = 2
+print(x.self) // prints 2
+```
+
+By analogy, we could spell the identity key path `\.self`, since it notionally
+refers to this `self` member:
+
+```swift
+let id = \Int.self
+
+x[keyPath: id] = 3
+print(x[keyPath: id]) // prints 3
+```
```

Comment:
It might be nice, to include in addition to this example, one that shows it being used on the ViewController example you motivated the proposal with.